lineapy/graph_reader/artifact_collection.py

```diff
@@ -42,55 +46,60 @@ def __init__(
+        check_duplicates(target_artifacts)
+        # Check if entries in reuse pre computed artifacts list have been repeated
+        check_duplicates(reuse_pre_computed_artifacts)
+
+        # ------ Initialize Linea Artifacts from artifactdef ------
+        # Retrieve target artifact objects
+        self.target_artifacts: List[
+            LineaArtifact
+        ] = get_db_artifacts_from_artifactdef(self.db, target_artifacts)
+
+        # Retrieve reuse precomputed artifact objects
+        self.pre_computed_artifacts: List[
+            LineaArtifact
+        ] = get_db_artifacts_from_artifactdef(
+            self.db, reuse_pre_computed_artifacts
         )
 
+        # ------ Group artifacts by session using SessionArtifact ------
         # For each session, construct SessionArtifacts object
         self.session_artifacts: Dict[LineaID, SessionArtifacts] = {}
         for (
             session_id,
-            session_artifacts,
-        ) in self.target_artifacts_by_session.items():
-            # TODO: LIN-653, LIN-640 only collect matched pre_computed artifacts in this Session
-            # instead of grabbing all Artifacts as currently required by SessionArtifact
-            pre_calculated_artifacts: List[LineaArtifact] = sum(
-                self.pre_computed_artifacts_by_session.values(), []
-            )
+            target_artifacts_by_session,
+        ) in get_artifacts_grouped_by_session(self.target_artifacts):
 
             self.session_artifacts[session_id] = SessionArtifacts(
                 self.db,
-                session_artifacts,
+                target_artifacts_by_session,
                 input_parameters=input_parameters,
-                reuse_pre_computed_artifacts=pre_calculated_artifacts,
+                # TODO: LIN-653, LIN-640 only collect matched pre_computed artifacts in this Session
+                # instead of grabbing all Artifacts as currently required by SessionArtifact
+                reuse_pre_computed_artifacts=self.pre_computed_artifacts,
             )
 
         # TODO: LIN-653, LIN-640
         # Add some type of validation that all re-use compute artifacts can be matched to
         # a target artifact before the creation of SessionArtifacts.
+        # If calls to .values() are repeated in the class, it might indicate a need for a new datastructure to hold the values
+        all_sessions_artifacts = [
+            art.name
+            for sa in self.session_artifacts.values()
+            for art in sa.all_session_artifacts.values()
+        ]
+
         # Check all reuse_pre_computed artifacts is used
-        all_sessions_artifacts = []
-        for sa in self.session_artifacts.values():
-            all_sessions_artifacts += [
-                art.name for art in sa.all_session_artifacts.values()
-            ]
-        pre_calculated_artifacts = sum(
-            self.pre_computed_artifacts_by_session.values(), []
```

Comment:
self.pre_computed_artifacts_by_session is generated from a list and grouped by session id. However, the dict form is never used, and for every use, the dict is flattened back to a list. An indication that grouping isn't needed here. Even if the grouped form is needed in the future, it can be held as a list and grouped on demand.